lib/tree.rb

```diff
@@ -2,59 +2,133 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+  end
+
+  def add_helper(current, new_node)
+    return new_node if current.nil?
+
+    if new_node.key <= current.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+    return current
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: worst case O(n)
+  # Space Complexity: worst case O(n)
   def find(key)
```

Comment:
👍 , Time complexity is O(log n) if the tree is balanced and O(n) if it's unbalanced.  Same for space complexity due to the system stack.